// homey-api/lib/HomeyAPI/HomeyAPIV2/App.js
'use strict';

const Item = require('./Item');

class App extends Item {

  async call({
    method,
    path,
    body,
  }) {
    if (typeof method !== 'string') {
      throw new Error('Invalid Path');
    }

    if (typeof path !== 'string') {
      throw new Error('Invalid Path');
    }

    if (!path.startsWith('/')) {
      path = `/${path}`;
    }

    this.__debug(method, path);

    return this.homey.call({
      method,
      path: `/api/app/${this.id}${path}`,
      body,
    });
  }

  /**
   * Make a GET request to the App's Web API.
   * @param {object} opts
   * @param {string} opts.path
   * @returns {Promise<any>}
   * @function HomeyAPIV2.ManagerApps.App#get
   */
  async get({ path }) {
    return this.call({
      path,
      method: 'GET',
    });
  }

  /**
   * Make a POST request to the App's Web API.
   * @param {object} opts
   * @param {string} opts.path
   * @param {object} opts.body
   * @returns {Promise<any>}
   * @function HomeyAPIV2.ManagerApps.App#post
   */
  async post({ path, body }) {
    return this.call({
      path,
      body,
      method: 'POST',
    });
  }

  /**
   * Make a PUT request to the App's Web API.
   * @param {object} opts
   * @param {string} opts.path
   * @param {object} opts.body
   * @returns {Promise<any>}
   * @function HomeyAPIV2.ManagerApps.App#put
   */
  async put({ path, body }) {
    return this.call({
      path,
      body,
      method: 'PUT',
    });
  }

  /**
   * Make a DELETE request to the App's Web API.
   * @param {object} opts
   * @param {string} opts.path
   * @returns {Promise<any>}
   * @function HomeyAPIV2.ManagerApps.App#delete
   */
  async delete({ path }) {
    return this.call({
      path,
      method: 'DELETE',
    });
  }

}

module.exports = App;
