// homey-api/lib/HomeyAPI/HomeyAPIApp.js
'use strict';

const HomeyAPI = require('./HomeyAPI');
const HomeyAPIV2 = require('./HomeyAPIV2');

/**
 * Use this class to utilize Homey Pro's Web API from inside an app.
 *
 * > This class only works on apps using Apps SDK v3 running on Homey Pro.
 *
 * > Make sure your app has the `homey:manager:api` permission.
 *
 * @class
 * @extends HomeyAPIV2
 * @example
 * // app.json
 * {
 *   ...
 *   "platforms": [ "local" ],
 *   "permissions": [ "homey:manager:api" ],
 * }
 *
 * // app.js
 * const Homey = require('homey');
 * const { HomeyAPIApp } = require('homey-api');
 *
 * class MyApp extends Homey.App {
 *
 *   async onInit() {
 *     const api = new HomeyAPIApp({
 *       homey: this.homey,
 *     });
 *
 *     const devices = await api.devices.getDevices();
 *     this.log('Devices:', devices);
 *   }
 *
 * }
 */
class HomeyAPIApp extends HomeyAPIV2 {

  /**
   * Possible Discovery Strategies
   * @static
   * @property {object} DISCOVERY_STRATEGIES
   * @property {string} DISCOVERY_STRATEGIES.LOCAL - Local HTTP, e.g. `http://192.168.1.100`.
   */
  static DISCOVERY_STRATEGIES = {};

  /**
   * @param {Homey} homey - The Homey instance of your app, usually `this.homey`.
   * @param {boolean} [debug=false] - Enable debug logs.
   */
  constructor({
    homey,
    debug = false,
    ...props
  }) {
    super({
      ...props,
      api: null,
      properties: {
        id: 'local',
        softwareVersion: null,
      },
      strategy: [
        HomeyAPI.DISCOVERY_STRATEGIES.LOCAL,
      ],
    });

    Object.defineProperty(this, '__homeySDK', {
      value: homey,
      enumerable: false,
      writable: true,
    });

    Object.defineProperty(this, '__debugEnabled', {
      value: !!debug,
      enumerable: false,
      writable: true,
    });
  }

  __debug(...props) {
    if (!this.__debugEnabled) return;
    this.__homeySDK.log('[homey-api]', `[${this.constructor.name}]`, ...props);
  }

  async login() {
    if (!this.__loginPromise) {
      this.__loginPromise = Promise.resolve().then(async () => {
        const token = await this.__homeySDK.api.getOwnerApiToken();
        this.__debug(`Local Token: ${token}`);
        return token;
      });

      this.__loginPromise
        .then(token => {
          this.__token = token;
        })
        .catch(err => {
          this.__debug('Error Logging In:', err);
        })
        .finally(() => {
          this.__loginPromise = null;
        });
    }

    return this.__loginPromise;
  }

  async discoverBaseUrl() {
    this.__baseUrl = await this.__homeySDK.api.getLocalUrl();
    this.__strategyId = 'local';

    this.__debug(`Local URL: ${this.__baseUrl}`);

    return {
      baseUrl: this.__baseUrl,
      strategyId: this.__strategyId,
    };
  }

}

module.exports = HomeyAPIApp;
