// homey-api/lib/HomeyAPI/HomeyAPI.js
'use strict';

/**
 * A Homey API, to be extended.
 * @hideconstructor
 */
class HomeyAPI {

  /**
   * Possible Discovery Strategies
   * @static
   * @property {object} DISCOVERY_STRATEGIES
   * @property {string} DISCOVERY_STRATEGIES.CLOUD - Cloud HTTPS, e.g. `https://abcdef.connect.athom.com`.
   * @property {string} DISCOVERY_STRATEGIES.MDNS - Local HTTP, e.g. `http://homey-abcdef.local`
   * @property {string} DISCOVERY_STRATEGIES.LOCAL - Local HTTP, e.g. `http://192.168.1.100`.
   * @property {string} DISCOVERY_STRATEGIES.LOCAL_SECURE - Local HTTPS, e.g. `https://192-168-1-100.homey.homeylocal.com`.
   * @property {string} DISCOVERY_STRATEGIES.REMOTE_FORWARDED - Remote HTTPS, e.g. `https://12-34-56-78.homey.homeylocal.com:12345`.
   */
  static DISCOVERY_STRATEGIES = {
    MDNS: 'mdns',
    CLOUD: 'cloud',
    LOCAL: 'local',
    LOCAL_SECURE: 'localSecure',
    REMOTE_FORWARDED: 'remoteForwarded',
  };

  constructor({
    api,
    properties,
    debug = false,
  }) {
    // Set Debug Enabled
    Object.defineProperty(this, '__debugEnabled', {
      value: !!debug,
      enumerable: false,
      writable: true,
    });

    // Set API
    Object.defineProperty(this, '__api', {
      value: api,
      enumerable: false,
      writable: true,
    });

    // Set ID
    Object.defineProperty(this, 'id', {
      value: properties._id || null,
      enumerable: true,
      writable: true,
    });

    // Set Version
    Object.defineProperty(this, 'version', {
      value: properties.softwareVersion || null,
      enumerable: true,
      writable: true,
    });

    // Set Properties
    Object.defineProperty(this, '__properties', {
      value: properties,
      enumerable: false,
      writable: false,
    });
  }

  __debug(...props) {
    this.__api.__debug(`[${this.constructor.name}:${this.id}]`, ...props);
  }

  /*
   * Storage
  */

  async __getStore() {
    if (!this.__api) return {};

    const store = await this.__api.__getStore();
    return store[`homey-${this.id}`] || {};
  }

  async __setStore(value) {
    if (!this.__api) return;

    const store = await this.__getStore();
    await this.__api.__setStore({
      [`homey-${this.id}`]: {
        ...store,
        ...value,
      },
    });
  }

  /*
   * Internationalization
   */

  /**
  * Translates an i18n-object (e.g. `{ en: 'My String', nl: 'Mijn tekst' }` to a string.
  * Uses the language of Homey as defined in {@link Homey}.
  * @param {object} input
  * @param {string} input.en - English translation
  * @param {string} input.nl - Dutch translation
  * @returns {string|null} - The translated string, or null
  * @example
  * homeyApi.__({
  *   en: 'Hello',
  *   nl: 'Hallo',
  *   fr: 'Bonjour',
  * }); // returns "Hello" if Homey is set to English
  */
  __(input) {
    if (typeof input === 'object' && input !== null) {
      if (typeof input[this.__properties.language] === 'string') {
        return input[this.__properties.language];
      }

      if (typeof input['en'] === 'string') {
        return input['en'];
      }
    }

    return null;
  }

}

module.exports = HomeyAPI;
