// homey-api/index.js
/* eslint-disable global-require */

'use strict';

module.exports = {
  AthomAppsAPI: require('./lib/AthomAppsAPI'),
  AthomBackupAPI: require('./lib/AthomBackupAPI'),
  AthomCallbackAPI: require('./lib/AthomCallbackAPI'),
  AthomCloudAPI: require('./lib/AthomCloudAPI'),
  AthomConnectAPI: require('./lib/AthomConnectAPI'),
  AthomDNSAPI: require('./lib/AthomDNSAPI'),
  AthomFirmwareAPI: require('./lib/AthomFirmwareAPI'),
  AthomFlowAPI: require('./lib/AthomFlowAPI'),
  AthomGeolocateAPI: require('./lib/AthomGeolocateAPI'),
  AthomIconsAPI: require('./lib/AthomIconsAPI'),
  AthomInfraredAPI: require('./lib/AthomInfraredAPI'),
  AthomNotificationsAPI: require('./lib/AthomNotificationsAPI'),
  AthomSetupAPI: require('./lib/AthomSetupAPI'),
  AthomStoreAPI: require('./lib/AthomStoreAPI'),
  AthomWeatherAPI: require('./lib/AthomWeatherAPI'),
  AthomWebhooksAPI: require('./lib/AthomWebhooksAPI'),
  HomeyCloudAPI: require('./lib/HomeyCloudAPI'),
  HomeyAPIApp: require('./lib/HomeyAPI/HomeyAPIApp'),
};
