// homey-api/lib/EventEmitter.js
'use strict';

/**
 * @class
 * @private
 */
class EventEmitter {

  constructor() {
    // Set Listeners
    Object.defineProperty(this, '__listeners', {
      value: {},
      enumerable: false,
      writable: false,
    });
  }

  /**
   * @param {string} event
   * @param  {...any} data
   */
  emit(event, ...data) {
    const listeners = this.__listeners[event];
    if (Array.isArray(listeners)) {
      listeners.forEach(listener => {
        listener(...data);
      });
    }
  }

  /**
   * @param {string} event
   * @param {function} callback
   */
  on(event, callback) {
    this.__debug(`.on('${event}')`);

    this.__listeners[event] = this.__listeners[event] || [];
    this.__listeners[event].push(callback);
  }

  /**
   * @param {string} event
   * @param {function} callback
   */
  once(event, callback) {
    this.__debug(`.once('${event}')`);

    const callback_ = (...args) => {
      this.__listeners[event] = this.__listeners[event].filter(cb => cb !== callback_);
      callback(...args);
    };

    this.__listeners[event] = this.__listeners[event] || [];
    this.__listeners[event].push(callback_);
  }

  /**
   * @param {string} event
   * @param {function} callback
   */
  off(event, callback) {
    this.__debug(`.off('${event}')`);

    this.__listeners[event] = this.__listeners[event].filter(cb => cb !== callback);
  }

  /**
   * Shortcut to {@link EventEmitter#off}
   * @param {string} event
   * @param {function} callback
   */
  removeListener(event, callback) {
    this.off(event, callback);
  }

  /**
   * @param {string} event
   */
  removeAllListeners(event) {
    this.__listeners[event] = [];
  }

}

module.exports = EventEmitter;
